test/integration/mongodb-handshake/mongodb-handshake.test.ts

```diff
@@ -0,0 +1,45 @@
+import { expect } from 'chai';
+import * as sinon from 'sinon';
+
+import {
+  Connection,
+  LEGACY_HELLO_COMMAND,
+  MongoServerError,
+  MongoServerSelectionError
+} from '../../mongodb';
+
+describe('MongoDB Handshake Node tests', () => {
+  let client;
+
+  context('when hello is too large', () => {
+    before(() => {
+      sinon.stub(Connection.prototype, 'command').callsFake(function (ns, cmd, options, callback) {
+        // @ts-expect-error: sinon will place wrappedMethod there
+        const command = Connection.prototype.command.wrappedMethod.bind(this);
+
+        if (cmd.hello || cmd[LEGACY_HELLO_COMMAND]) {
+          return command(
+            ns,
+            { ...cmd, client: { driver: { name: 'a'.repeat(1000) } } },
+            options,
+            callback
+          );
+        }
+        return command(ns, cmd, options, callback);
+      });
+    });
+
+    after(() => sinon.restore());
+
+    it('client fails to connect with an error relating to size', async function () {
+      client = this.configuration.newClient({ serverSelectionTimeoutMS: 2000 });
+      const error = await client.connect().catch(error => error);
+      if (this.configuration.isLoadBalanced) {
+        expect(error).to.be.instanceOf(MongoServerError);
```

Comment:
Interesting that LB has this behavior, but I guess it makes sense

Reply:
Just a note on the comment - load balancer mode has its own custom server selector that just grabs the same server every time (the only server in the topology) so it never gets actual client side server selection errors.